// cards/cards.js
//Cards are responsive as long as their height:width ratio stays 10:6
//Card type pictures need to be squares but can be swapped out

function Card(id, user_card, rotate){
  this.elem = buildCard(id, user_card, rotate);
  this.suit = null;
  this.val  = null;
}

function buildCard(id, user_card, rotate){
  var cardContainer;

  cardContainer = $('<div/>', {
    id: id,
    class: 'card' + (rotate ? '-rotate' : '')
  });

  cardContainer.hover(
    function(){
      var t = parseInt(this.style.top) - 30;
      this.style.top = t + 'px';
    },
    function(){
      var t = parseInt(this.style.top) + 30;
      this.style.top = t + 'px';
    });

  if(user_card){
    cardContainer = cardContainer[0];
    cardContainer.appendChild(buildTopContainer());
    cardContainer.appendChild(buildMidContainer());
    cardContainer.appendChild(buildBtmContainer());
  } else {
    cardContainer.css({ 'background-color' : 'red' });
    cardContainer = cardContainer[0];
  }

  return cardContainer;
}

function buildTopContainer(){
  var topContainer, type;

  topContainer = $('<div/>', {
    class: 'card-top-container'
  })[0];

  type = $('<img/>', {
    class: 'small-card-type abs-left',
    src: 'prod/images/default_card_type.png'
  })[0];

  topContainer.appendChild(type);

  return topContainer;
}

function buildMidContainer(){
  var midContainer, type;

  midContainer = $('<div/>', {
    class: 'card-mid-container'
  })[0];

  type = $('<img/>', {
    class: 'large-card-type',
    src: 'prod/images/default_card_type.png'
  })[0];

  midContainer.appendChild(type);

  return midContainer;
}

function buildBtmContainer(){
  var btmContainer, type;

  btmContainer = $('<div/>', {
    class: 'card-btm-container'
  })[0];

  type = $('<img/>', {
    class: 'small-card-type abs-right',
    src: 'prod/images/default_card_type.png'
  })[0];

  btmContainer.appendChild(type);

  return btmContainer;
}

module.exports = Card;
