// cards/cards.js
//Cards are responsive as long as their height:width ratio stays 10:6
//Card type pictures need to be squares but can be swapped out

function Card(id, user_card, rotate, val, src){
  this.user_card = user_card;
  this.value  = val;
  this.src = src;
  this.elem = this.buildCard(id, user_card, rotate);
}

Card.prototype.show = function() {
  if(!this.user_card){
    $(this.elem).css({'background-color' : 'white'});
    this.elem.appendChild(this.buildTopContainer());
    this.elem.appendChild(this.buildMidContainer());
    this.elem.appendChild(this.buildBtmContainer());
  }
};

Card.prototype.unshow = function() {
  $(this.elem).empty();
  $(this.elem).css({'background-color' : 'red'});
};

Card.prototype.buildCard = function(id, user_card, rotate){
  var cardContainer;

  cardContainer = $('<div/>', {
    id: id,
    class: 'card' + (rotate ? '-rotate' : '')
  });

  if(user_card){

    cardContainer.hover(
    function(){
      var t = parseInt(this.style.top) - 30;
      this.style.top = t + 'px';
    },
    function(){
      var t = parseInt(this.style.top) + 30;
      this.style.top = t + 'px';
    });  

    cardContainer = cardContainer[0];
    cardContainer.appendChild(this.buildTopContainer());
    cardContainer.appendChild(this.buildMidContainer());
    cardContainer.appendChild(this.buildBtmContainer());
  } else {
    cardContainer.css({ 'background-color' : 'red' });
    cardContainer = cardContainer[0];
  }

  return cardContainer;
}

Card.prototype.buildTopContainer = function(){
  var topContainer, type;

  topContainer = $('<div/>', {
    class: 'card-top-container'
  })[0];

  type = $('<img/>', {
    class: 'small-card-type abs-left',
    src: this.src
  })[0];

  topContainer.appendChild(type);

  return topContainer;
}

Card.prototype.buildMidContainer = function(){
  var midContainer, type;

  midContainer = $('<div/>', {
    class: 'card-mid-container'
  })[0];

  type = $('<img/>', {
    class: 'large-card-type',
    src: this.src
  })[0];

  midContainer.appendChild(type);

  return midContainer;
}

Card.prototype.buildBtmContainer = function(){
  var btmContainer, type;

  btmContainer = $('<div/>', {
    class: 'card-btm-container'
  })[0];

  type = $('<img/>', {
    class: 'small-card-type abs-right',
    src: this.src
  })[0];

  btmContainer.appendChild(type);

  return btmContainer;
}

module.exports = Card;
