// ai/ai.js

//PROTOTYPAL ADDS
//==============================================================================

Object.defineProperty( Array.prototype, 'last', {
  value: function(){
    return this.length > 0 ? this[this.length - 1] : null; 
  },
  enumerable: false
});

Object.defineProperty( Array.prototype, 'getRandom', {
  value: function(prev_idxs){
    var min, max, rand;

    if(this.length == 0 || (prev_idxs && prev_idxs.length == this.length)){
        return;
    }

    min = 0;
    max = this.length;
    rand = Math.floor(Math.random() * (max - min)) + min;

    if(!prev_idxs){
      return this[rand];
    }

    while(prev_idxs.contains(rand)){
        rand = Math.floor(Math.random() * (max - min)) + min;
    }

    prev_idxs.push(rand);
    return this[rand];
  },
  enumerable: false
});

Object.defineProperty( Array.prototype, 'getUniqueRandomValues', {
  value: function(num_vals){
    var prev, vals;

    prev = [];
    vals = [];

    for(var i = 0; i < num_vals; i++){
        vals.push(this.getRandom(prev));
    }

    return vals;
  },
  enumerable: false
});

Object.defineProperty( Array.prototype, 'getRandomValues', {
  value: function(num_vals){
    var vals;

    vals = [];

    for(var i = 0; i < num_vals; i++){
        vals.push(this.getRandom());
    }

    return vals;
  },
  enumerable: false
});

Object.defineProperty( Array.prototype, 'contains', {
  value: function(obj){
    return this.indexOf(obj) != -1;
  },
  enumerable: false
});

//GAME CLASSES
//==============================================================================
var Hand = require('hand');


var AI = function(player){
    var players = [];
    players.push(player);
    players.push(new Player('player 1'));
    players.push(new Player('player 2'));
    players.push(new Player('player 3'));

    this.level = 0;
    this.played_turns = 0;
    this.game = new Game(this.level, players, this);

    this.nextRound();
}

AI.prototype.nextGame = function() {
    this.level++;
    this.game = new Game(this.level, this.game.players);
};

AI.prototype.nextRound = function() {
    if(this.game.curr_round){
        this.game.addRound(this.game.curr_round);
    }

    this.game.rounds.push(new Round());
};

AI.prototype.playCard = function(player, card) {
    this.game.rounds.last().playTurn(player, card);

    if(this.played_turns != 3){
        this.played_turns++;
    } else {
        this.played_turns = 0;
        console.log(this.pickRoundWinner());
    }
};

AI.prototype.pickRoundWinner = function() {
    var winner = null;

    this.game.round_rules.forEach(function(rule){
        winner = (rule.pickWinner(this.game) || winner);
    }.bind(this));

    this.game.rounds.last().winner = winner;
    return winner;
};

AI.prototype.pickGameWinner = function() {
    return this.game.hand_rule.pickWinner(this.game);
};


/**
 * A Game holds all of the state for the current game being played. This includes
 * the rules of the game and the rounds that have been played in order.
 * @param {int} level 
 */
var Game = function(level, players, ai){
    this.round_rules = this.getRoundRules(level + 3);
    this.hand_rule = this.getHandRule();
    this.rounds = [];
    this.players = players;
    this.curr_round = null;

    this.newHand(ai);
}

Game.prototype.getRoundRules = function(num_rules) {
    return ROUNDRULES;//.getUniqueRandomValues(num_rules);
};

Game.prototype.getHandRule = function() {
    return HANDRULES[0];//.getRandom([]);
};

Game.prototype.addRound = function(round) {
    this.rounds.push(round);
};

Game.prototype.newHand = function(ai) {
    var btm, top,lft, rght;

    btm = new Hand(13, {
      user_card: true,
      rotate: false,
      position: 'bottom',
      player: this.players[0],
      ai: ai
    }).show($("#game-screen-bottom")[0], 50);

    lft = new Hand(13, {
      user_card: false,
      rotate: true,
      position: 'left',
      player: this.players[1],
      ai: ai
    }).show($('#game-screen-side-left')[0], 50);

    top = new Hand(13, {
      user_card: false,
      rotate: false,
      position: 'top',
      player: this.players[2],
      ai: ai
    }).show($("#game-screen-top")[0], 50);

    rght = new Hand(13, {
      user_card: false,
      rotate: true,
      position: 'right',
      player: this.players[3],
      ai: ai
    }).show($('#game-screen-side-right')[0], 50);
};


/**
 * Round holds the information for each round of play. A round ends once every
 * player has played a card.
 */
var Round = function(){
    this.turns = [];
    this.winner = null;
}

Round.prototype.playTurn = function(player, card) {
    this.turns.push({ 'player' : player, 'card': card });
};



/**
 * Rule to determine who won either a hand or a round
 * @param {function} solver function that takes a game and determines the winner 
 * of either the current round or of the whole hand
 */
var Rule = function(solver){
    this.solver = solver
}

Rule.prototype.pickWinner = function(game) {
    return this.solver(game);
};


//GAME CONSTANTS
//==============================================================================
function highCardWins(game){
    var round, winner;

    winner = null;
    round = game.rounds.last();
    
    round.turns.forEach(function(turn){
        if(!winner){
            winner = turn;
        }

        if(winner.card.value < turn.card.value){
            winner = turn;
        }
    });

    return winner;
}

function eightsAreWild(game){
    var round, winner;

    winner = null;
    round = game.rounds.last();

    round.turns.forEach(function(turn){
        if(turn.card.value == 8 && (!winner || winner.card.value != 8)){
            winner = turn;
        }
    });

    return winner;
}

var ROUNDRULES = [
    new Rule(highCardWins),
    new Rule(eightsAreWild)
]

function MostRounds(game){
    var rounds, winner;

    rounds = game.rounds;

    winner = rounds[0].winner
    rounds.forEach(function(round){
        round.winner.player.rounds_won++

        if(winner.player.rounds_won < round.winner.player.rounds_won){
            winner = round.winner;
        }
    });

    return winner.player;
}

var HANDRULES = [
    new Rule(MostRounds)
]

module.exports = AI;


var Player = function(name){
    this.rounds_won = 0;
    this.name = name;
    this.is_turn = false;
}
